### PR TITLE
[CS2113T-T12-4] PAC

### PR DESCRIPTION
**Pull request by lowjiayee**
@lowjiayee
@harithadiv
@benchan911
@anqi-nus

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -1,34 +1,179 @@
...
+### Event
+![event](images/event.png "Class diagram of Event component")           
+*Class diagram of the Event component*
+
+1. When a user enters an event-related command, the command is analysed by `EventCommandInterpreter`. 
+1. Once determined, the relevant information (e.g. index, name, time, date, venue) are extracted by 
+`EventParser`.
+1. Then, the relevant class that corresponds to the command is created, with the information extracted 
+from the previous step passed into it. It modifies `Event` or `EventList`.
+1. These commands are then returned to `Duke.run()` to `execute()`.
```
Comment (JensonWee) [inline docs/DeveloperGuide.md]: The diagram looks neat and very informative. Perhaps more description would aid the user a better understanding of the diagram.

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -1,34 +1,179 @@
...
+* Any classes (e.g. `Seminar`) that inherit from `Event` class will have similar control flow. 
+
+### Attendance
+![attendance](images/Attendance.png "Class diagram of Attendance component")        
+*Class diagram of the Attendance component*
+1. When a user enters an attendance-related command, the command is analysed by `AttendanceCommandInterpreter`. 
+1. Once determined, the relevant class that corresponds to the type of command is created.
+1. Then, the class will execute base on its function. It modifies `AttendanceList`.
+1. These commands are then returned to `Duke.run()` to `execute()`. 
+
```
Comment (JensonWee) [inline docs/DeveloperGuide.md]: The diagram shows clearly what is been created when the function is executed. However, description of the attendance class is not detailed enough. Users might wonder where is AttendanceList? Perhaps a few sentence regarding the details of the objects will aid the user a better understanding of the class. Furthermore, what does the grey box represents?

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -1,34 +1,179 @@
...
+###Calendar Implementation
+
+*Figure 2: Class diagram of the Calendar component*
+
```
Comment (JensonWee) [inline docs/DeveloperGuide.md:175]: Class diagram of the Calendar component is missing from DG. Is this going to be added in the future update? Perhaps telling the user that it would be added in the future would let the user know that it is going to be added.

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -1,34 +1,179 @@
...
+* `datetime` is stored as a single attribute in `Event` class, but it is exposed to user as `date` 
+and `time`, which corresponds to `d/` and `t/` flag respectively.
+* `editDate` or `editTime` commands are not available. Only `editDateTime` is available to change the 
+`date` and/or `time` of an `Event` object.
+* `delete(Event)` method is currently not in use, but can be used to implement delete by event name, 
+either by complete match, or fuzzy match.
+* Any classes (e.g. `Seminar`) that inherit from `Event` class will have similar control flow. 
+
+### Attendance
+![attendance](images/Attendance.png "Class diagram of Attendance component")
```
Comment (JensonWee) [inline docs/DeveloperGuide.md]: The diagram does not display AttendanceList. Is this format correct? Perhaps an indication where AttendanceList is from would aid the user better understanding of the diagram.

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -1,34 +1,179 @@
...
+1. Once determined, the relevant class that corresponds to the type of command is created.
+1. Then, the class will execute base on its function. It modifies `AttendanceList`.
+1. These commands are then returned to `Duke.run()` to `execute()`. 
+
+###Calendar Implementation
+
+*Figure 2: Class diagram of the Calendar component*
+
+1. When a user enters a calendar-related command, the command is analysed by `CalendarCommandInterpreter`.
+1. Once determined, the relevant information (eg. semester, academic year) are extracted by `CalendarParser`.
```
Comment (JensonWee) [inline docs/DeveloperGuide.md:177]: Perhaps clearer description of what's 'relevant information' would aid the user in better understanding. Is semester and academic year the only thing that is extracted by the `CalenderParser`?

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -1,34 +1,179 @@
...
+`date` and/or `time` of an `Event` object.
+* `delete(Event)` method is currently not in use, but can be used to implement delete by event name, 
+either by complete match, or fuzzy match.
+* Any classes (e.g. `Seminar`) that inherit from `Event` class will have similar control flow. 
+
+### Attendance
+![attendance](images/Attendance.png "Class diagram of Attendance component")        
+*Class diagram of the Attendance component*
+1. When a user enters an attendance-related command, the command is analysed by `AttendanceCommandInterpreter`. 
+1. Once determined, the relevant class that corresponds to the type of command is created.
```
Comment (JensonWee) [inline docs/DeveloperGuide.md]: Perhaps clearer description of what's 'relevantclass' would aid the user in better understanding.

File: docs/DeveloperGuide.md (hunk 4 above)
Comment (JensonWee) [reply docs/DeveloperGuide.md]: Perhaps a sequence diagram will help users to understand the flow of the command execution?

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -1,34 +1,179 @@
...
+###Calendar Implementation
+
+*Figure 2: Class diagram of the Calendar component*
+
+1. When a user enters a calendar-related command, the command is analysed by `CalendarCommandInterpreter`.
+1. Once determined, the relevant information (eg. semester, academic year) are extracted by `CalendarParser`.
+1. Then, either AddFirstSemester or AddSecondSemester class that corresponds the semester number is created. 
+1. Subsequently, it separates events by the required month and year in `CalendarList`
+1. These commands are then returned to `Duke.run()` to `execute()`. 
+
```
Comment (JensonWee) [inline docs/DeveloperGuide.md]: Description of the flow could be more detailed. Perhaps you can include the reason why such methods are implemented as such. The Pros and Cons of the methods and alternative methods.

File: docs/DeveloperGuide.md (hunk 2 above)
Comment (JensonWee) [reply docs/DeveloperGuide.md]: Description of the flow could be more detailed. Perhaps you can include the reason why such methods are implemented as such. The Pros and Cons of the methods and alternative methods.

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -1,34 +1,179 @@
...
+## Architecture
+This section presents the architecture of PAC. It explains the architecture of main components of PAC.
+
+### Overall Architecture
+*Overall Class diagram*
+{To be added in future revisions}
+
+### Command
+![Command](images/Command.png "Class diagram of Command component")
```
Comment (andy-aw-why) [inline docs/DeveloperGuide.md:91]: Is it possible to resize or restructure this diagram such that it is bigger?

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -1,34 +1,179 @@
...
+## Architecture
+This section presents the architecture of PAC. It explains the architecture of main components of PAC.
+
+### Overall Architecture
+*Overall Class diagram*
+{To be added in future revisions}
+
+### Command
+![Command](images/Command.png "Class diagram of Command component")         
+*Class diagram of the Command component*
```
Comment (thanhduc2000) [inline docs/DeveloperGuide.md:92]: The diagram is hard to follow since the image is compressed and not viewable in the website and also not zoomable. Moreover, you are using an auto generated diagram from the system, which is not recommended

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -1,34 +1,179 @@
...
 
-{Describe the design and implementation of the product. Use UML diagrams and short code snippets where applicable.}
+UI is the main class handles user display, which includes reading user input and printing information 
+back to the user on command-line.  
+Besides the normal command line messages, there are two subclasses of UI: 
+DisplayList and DisplayTable, to specifically print the list and table interface to professor. 
 
+## Feature Design and Implementation 
+### Event
+![event](images/event.png "Class diagram of Event component")
```
Comment (andy-aw-why) [inline docs/DeveloperGuide.md]: Hi, is the format for this diagram (and other diagrams as well in this guide) correct? I think it is missing the private/public field for variable and method name.

Comment (andy-aw-why) [reply docs/DeveloperGuide.md]: Also, is this generated using SimpleUML plugin for intellij? I don't think it is recommended to auto generate the UML diagrams.

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -1,34 +1,179 @@
...
-{Describe the design and implementation of the product. Use UML diagrams and short code snippets where applicable.}
+UI is the main class handles user display, which includes reading user input and printing information 
+back to the user on command-line.  
+Besides the normal command line messages, there are two subclasses of UI: 
+DisplayList and DisplayTable, to specifically print the list and table interface to professor. 
 
+## Feature Design and Implementation 
+### Event
+![event](images/event.png "Class diagram of Event component")           
+*Class diagram of the Event component*
```
Comment (thanhduc2000) [inline docs/DeveloperGuide.md]: This diagram is informative and easier to follow. However, since this is auto-generated, there are some aspects that auto generated diagrams cannot express some terms in words like accessibility in methods (public, private, etc), and they refer to icons instead

File: docs/DeveloperGuide.md (hunk 4 above)
Comment (Shannonwje) [inline docs/DeveloperGuide.md]: AttendanceList can't be seen inside the diagram, is this format correct? Maybe including AttendanceList would make understanding the diagram clearer, it can be in another diagram if AttendanceList can't fit into the current diagram or make the current diagram less neat.

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -1,34 +1,179 @@
...
+
+1. When a user enters an event-related command, the command is analysed by `EventCommandInterpreter`. 
+1. Once determined, the relevant information (e.g. index, name, time, date, venue) are extracted by 
+`EventParser`.
+1. Then, the relevant class that corresponds to the command is created, with the information extracted
```
Comment (Shannonwje) [inline docs/DeveloperGuide.md]: Diagram and explanation of the overall design and implementation looks neat but maybe giving an example of what you mean by relevant classes would make the understanding of the diagram better.

File: docs/DeveloperGuide.md (hunk 4 above)
Comment (thanhduc2000) [inline docs/DeveloperGuide.md]: Should add AttendanceList alo since it is covered in ListAttendance

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -1,34 +1,179 @@
...
+
+### Attendance
+![attendance](images/Attendance.png "Class diagram of Attendance component")        
+*Class diagram of the Attendance component*
+1. When a user enters an attendance-related command, the command is analysed by `AttendanceCommandInterpreter`. 
+1. Once determined, the relevant class that corresponds to the type of command is created.
+1. Then, the class will execute base on its function. It modifies `AttendanceList`.
+1. These commands are then returned to `Duke.run()` to `execute()`. 
+
+###Calendar Implementation
```
Comment (Shannonwje) [inline docs/DeveloperGuide.md]: The header for Calendar Implementation seems to be off, it appears as "###Calendar Implementation" rather than an actual header of the section.

Comment (Shannonwje) [reply docs/DeveloperGuide.md]: One change would be to make it "### Calendar Implementation" with a space in between instead.

Comment (Shannonwje) [reply docs/DeveloperGuide.md]: Another suggestion would be to be consistent with the header names, to be "### Calandar" to be consistent with the header names before like "### Event" and "### Attendance"

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -1,34 +1,179 @@
...
+
+*Figure 2: Class diagram of the Calendar component*
```
Comment (Shannonwje) [inline docs/DeveloperGuide.md]: The class diagram for the calendar component seems to be missing, will it be added in the  future? If it is currently not ready then maybe can consider not referencing to it now to avoid confusion for the readers.

**Review by thanhduc2000 — COMMENTED** [review]
Overall, some class diagrams can be improved. All diagrams should not be auto-generated and your team could use drawing tool online instead. Moreover, you could start working on sequential diagrams for easier explanation

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -1,34 +1,179 @@
...
+1. Once determined, the relevant information (eg. semester, academic year) are extracted by `CalendarParser`.
+1. Then, either AddFirstSemester or AddSecondSemester class that corresponds the semester number is created. 
+1. Subsequently, it separates events by the required month and year in `CalendarList`
+1. These commands are then returned to `Duke.run()` to `execute()`. 
+
+Note that:
+* `acadamic year` is parsed into corresponding to only one year according to the semester in `EventParser` class.
+* Calendar view of the whole year is not available. Only semester 1 or 2 of an academic year can be viewed at a time.
+* Event name size must be less than 10 characters to be displayed neatly (current implementation), however
+it can be implement to truncate longer names to fit nicely
```
Comment (Shannonwje) [inline docs/DeveloperGuide.md]: Overall Design and Implementation is neat and relatively easy to understand. Perhaps can consider adding in examples with sample inputs of how each implementation would work rather than simply explaining the flow.

Comment (Shannonwje) [reply docs/DeveloperGuide.md]: Another suggestion to help the reader understand more about your choice of design and implementation would be stating the different methods of implementation, weighing the pros and cons and stating the reason for choosing a certain implementation over another.

File: docs/DeveloperGuide.md (hunk 4 above)
Comment (Shannonwje) [reply docs/DeveloperGuide.md]: Also, there is a grey box in some of the classes while it is not in some other classes. Is there some information missing or omitted in the diagram? Perhaps an explanation for the grey boxes would make it less confusing for the readers trying to understand the diagram

**Review by Shannonwje — COMMENTED** [review]
Overall DG is clear and well-explained. Some minor errors in formatting of the headers and a missing diagram but overall easy to understand.

File: docs/DeveloperGuide.md (hunk 9 above)
Comment (SibingWu) [inline docs/DeveloperGuide.md]: I suppose it would be better if the diagram is not so clustered? Also it seems that the IDE 0generated diagram is not encouraged for project submission

File: docs/DeveloperGuide.md (hunk 13 above)
Comment (SibingWu) [inline docs/DeveloperGuide.md]: image
It seems that the title here is not working well. Maybe consider to add one more space before "Calender Implementation"? i.e. `### Calendar Implementation` instead of "###Calendar Implementation"

File: docs/DeveloperGuide.md (hunk 8 above)
Comment (sliu107) [reply docs/DeveloperGuide.md:91]: Is it possible to reorder the commands in the diagram so that the commands corresponding the same feature can be put together which makes it easier for the readers to understand?

File: docs/DeveloperGuide.md (hunk 4 above)
Comment (SibingWu) [inline docs/DeveloperGuide.md]: It seems that `AttendanceList` and `EventList` are missing in this diagram?

File: docs/DeveloperGuide.md (hunk 2 above)
Comment (SibingWu) [inline docs/DeveloperGuide.md]: Maybe you can add a sequence diagram here to illustrate the execution?

File: docs/DeveloperGuide.md (hunk 8 above)
Comment (sliu107) [reply docs/DeveloperGuide.md:91]: And I think omitting the constructors or some less important methods can simplify the commands diagram. In this way, it will be more easier for you to resize or restructure the diagram.

File: docs/DeveloperGuide.md [hunk 16]
```diff
@@ -1,34 +1,179 @@
...
+![Student](images/Student.png "Class diagram of Student component")     
+*Class diagram of the Student component*  
+1. When a user enters an studentList-related command, the command is analysed by `StudentCommandInterpreter`. 
+1. Once determined, the relevant class that corresponds to the type of command is created.
+1. Then, the class will execute base on its function. It modifies `AttendanceList`.
+1. These commands are then returned to `Duke.run()` to `execute()`. 
 
-{Describe the value proposition: what problem does it solve?}
+Note that:
+* studentList-related commands can be executed without the existence of events.
```
Comment (SibingWu) [inline docs/DeveloperGuide.md]: I am not sure if you format this on purpose, but it seems that those parts look similar to `Event`, `Attendance`, and `Calendar Implementation`. Maybe these two parts also belong to `Feature Design and Implementation` section?

File: docs/DeveloperGuide.md [hunk 17]
```diff
@@ -1,34 +1,179 @@
...
+![Command](images/Command.png "Class diagram of Command component")         
+*Class diagram of the Command component*  
+
+Commands are the main classes to be executed in PAC. All of the specific Command classes inherit the 
+base Command abstract class, and utilize its abstract execute() method.  
+A subclass of Command is created and executed when the professor input a corresponding command.
+ 
+### Parser
+*Class diagram of the Parser component*  
+There are total of four Parser classes as shown below. Each Parser class correspond to a feature
```
Comment (sliu107) [inline docs/DeveloperGuide.md:102]: It seems the diagram of parser compont is missing here.If you are going to add it in the feature, I think you can draw a class diagram here and add a sequence diagram to give an example at the end of this part.

File: docs/DeveloperGuide.md (hunk 4 above)
Comment (SibingWu) [reply docs/DeveloperGuide.md]: Consulted the prof: "It's OK to omit less important classes from the diagram"

File: docs/DeveloperGuide.md [hunk 18]
```diff
@@ -1,34 +1,179 @@
...
+1. Then, the class will execute base on its function. It modifies `AttendanceList`.
+1. These commands are then returned to `Duke.run()` to `execute()`. 
+
+###Calendar Implementation
+
+*Figure 2: Class diagram of the Calendar component*
+
+1. When a user enters a calendar-related command, the command is analysed by `CalendarCommandInterpreter`.
+1. Once determined, the relevant information (eg. semester, academic year) are extracted by `CalendarParser`.
+1. Then, either AddFirstSemester or AddSecondSemester class that corresponds the semester number is created.
```
Comment (SibingWu) [inline docs/DeveloperGuide.md]: Just minor format thing: maybe better to emphasize the two classes? `AddFirstSemester` or `AddSecondSemester`

**Review by SibingWu — COMMENTED** [review]
Structures are generally good and clear. It would be better if you can add more sequence diagrams and replace the IDE-generated diagram

File: docs/DeveloperGuide.md [hunk 19]
```diff
@@ -1,34 +1,179 @@
...
+| CalenderParser            | EventCommandInterpreter                                       | 
+| AttendanceParser          | Step-by-step command at Attendance-related command classes    |
+| PerformanceParser         | Step-by-step command at performance-related command classes   |  
+
+A Parser class is created when a user input contains data to be stored or used in certain features.    
 
-## Design & Implementation
+### UI
+![Ui](images/Ui.png "Class diagram of Ui component")                
+*Class diagram of the UI component*
```
Comment (sliu107) [inline docs/DeveloperGuide.md]: Is it possible to show the high level association between UI and command class in the diagram? Or just add some description in words so that the reader can more easily understand how the ui is used in other classes and how ui be implemented.

**Review by sliu107 — COMMENTED** [review]
All the diagrams are class diagrams, I think adding some sequence diagrams is helpful for readers to understand how each part work and how they are connected to each other.

File: docs/DeveloperGuide.md (hunk 8 above)
Comment (ananda-lye) [reply docs/DeveloperGuide.md:91]: May be better to redraw the commands separately

File: docs/DeveloperGuide.md (hunk 10 above)
Comment (ananda-lye) [reply docs/DeveloperGuide.md]: Might be better to redraw with the notations that we are more familiar with. Currently there are many arrows, some may be combined to reduce clutter.

File: docs/DeveloperGuide.md (hunk 4 above)
Comment (ananda-lye) [reply docs/DeveloperGuide.md]: Missing the AttendanceList class in the diagram. Agrees with previous comments about possibly adding a sequence diagram. Also, I feel that some arrows can be omitted.

File: docs/DeveloperGuide.md (hunk 12 above)
Comment (ananda-lye) [reply docs/DeveloperGuide.md]: Unsure of what the relevant class refers to. Can be phrased more explicitly maybe.

File: docs/DeveloperGuide.md (hunk 13 above)
Comment (ananda-lye) [inline docs/DeveloperGuide.md]: Seems to have some formatting error

**Review by ananda-lye — COMMENTED** [review]
Maybe the generated diagrams can be used as reference only. As other reviewers have pointed out, the notation may not be entirely correct.  Also, consider adding some sequence diagrams which will greatly aid in understanding how the classes interact with one another. Cheers ! :)
